### PR TITLE
new Dockerfile, new cache

### PR DESCRIPTION
**Pull request by TriplEight**
- new Dockerfile: https://github.com/paritytech/scripts/pull/63
- new cache is being saved in docker volume:
  - `CARGO_HOME` set for every job's name to avoid data race condition while `cargo` tries to access the cache. `Cargo` saves dependencies and their git repositories to `CARGO_HOME`. Saving this means it's no more needed to download dependencies.
  - `sccache` saves compilation cache and significantly increases compilation speed. Avoids race conditions by design (hopefully).
  `GIT_STRATEGY` allows GitLab to save repository in the docker cache and update it via `git fetch`

It's required to modify runners config.
Example:
```
[[runners]]
  name = "QA_experiments"
  url = "https://gitlab.parity.io"
  token = ""
  executor = "docker"
  output_limit = 16384
  [runners.docker]
    privileged = false
    security_opt = ["seccomp=unconfined"]
    disable_entrypoint_overwrite = false
    disable_cache = false
    volumes = ["/home/gitlab-runner/ci-cache:/ci-cache:rw"]
    shm_size = 0
  [runners.cache]
    [runners.cache.s3]
    [runners.cache.gcs]

```

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -106,12 +108,14 @@ test-linux-stable:                 &test
     variables:
       - $DEPLOY_TAG
   before_script:
-   - test -d ${CARGO_HOME} -a -d ./target &&
-     echo "build cache size:" &&
-     du -h --max-depth=2 ${CARGO_HOME} ./target
-   - ./scripts/build.sh
+    - test -d ${CARGO_HOME} -a -d ./target &&
```
Comment (DemiMarie-parity) [inline .gitlab-ci.yml]:
```suggestion
    - test -d "${CARGO_HOME}" -a -d ./target &&
```

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -106,12 +108,14 @@ test-linux-stable:                 &test
...
+      echo "build cache size:" &&
+      du -h --max-depth=2 ${CARGO_HOME} ./target
```
Comment (DemiMarie-parity) [inline .gitlab-ci.yml]:
```suggestion
      du -h --max-depth=2 "${CARGO_HOME}" ./target
```

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -158,6 +159,7 @@ build-linux-release:               &build
     - echo "\n# building node-template\n"
```
Comment (DemiMarie-parity) [inline .gitlab-ci.yml]:
```suggestion
    - printf '\n# building node-template\n\n'
```

**Comment by bkchr** [comment]
@TriplEight I would like to merge this, after 1.0 is branched. Will this work out of the box or will it require some other modifications to the CI as well?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -170,185 +164,188 @@ build-rust-doc-release:            &build
...
-deploy-ew3-tag:
-  <<:                              *deploy-tag
-  environment:
-    name: parity-prod-ew3
-
-deploy-ue1-tag:
-  <<:                              *deploy-tag
-  environment:
-    name: parity-prod-ue1
+# .publish-build:                    &publish-build
```
Comment (bkchr) [inline .gitlab-ci.yml]: If all this is not required anymore, please remove it.

**Comment by TriplEight** [comment]
@bkchr everything's ready for merging.
- new docker image is already on [hub.docker.com](https://cloud.docker.com/u/parity/repository/docker/parity/rust-substrate-build)
- `linux-docker` tagged runners are set up to handle cache from both Substrate and Parity-Ethereum
- latest run is [here](https://gitlab.parity.io/parity/substrate/pipelines/34715).

**Pull request merged**